### DISCUSSION TITLE
Setting multiple cookies in _middleware

### DISCUSSION THREAD

**Discussion by Thamasky** [vercel/community#259] [Help] [answered]
Hi all,

I'm having issues trying to set multiple cookies in the _middleware function. It works perfect on localhost, but when deployed on vercel it only sets one cookie. It could be due the fact that Vercel combines everything into one single set-cookie header.

Code that sets the cookies:
```
   res.cookie(CookieNames.VendorCode, vendorCode, {
      maxAge: 365 * 24 * 60 * 60,
      secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
      sameSite: 'none',
    });

    res.cookie(CookieNames.CountryCode, countryCode, {
      maxAge: 365 * 24 * 60 * 60,
      secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
      sameSite: 'none',
    });

```

I tried changing this behaviour by changing the `res.cookie()` statements to serializing the cookies first, and then appending them all to the headers like this:

```
   const cookies: string[] = [];

    cookies.push(
      serialize(CookieNames.VendorCode, vendorCode, {
        maxAge: 365 * 24 * 60 * 60,
        secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
        sameSite: 'none',
      }),
    );

    cookies.push(
      serialize(CookieNames.CountryCode, countryCode, {
        maxAge: 365 * 24 * 60 * 60,
        secure: process.env.NEXT_PUBLIC_VERCEL_ENV !== 'development',
        sameSite: 'none',
      }),
    );

    cookies.forEach((cookie) => {
      res.headers.append('Set-Cookie', cookie);
    });
```

This sets multiple set-cookie headers on localhost, but when deployed to Vercel it gets reduced to a single set-cookie header again with all cookies in it. However it only sets one cookie, so the problem persists.

Example set-cookie header on Vercel:

```
set-cookie: ec-vc=ec; Max-Age=31536000; Secure; SameSite=None, ec-cc=be; Max-Age=31536000; Secure; SameSite=None
```

I also already tried to add domain & path to the cookies. But no difference there.

Any help on this would be really appreciated.

Kind regards,
Thomas

**Comment by CatsAreMLG** [comment]
Hi there!

Through this channel, we can only assist with questions and issues related to the Vercel platform and are unable to provide code-level support.

I would reach out to our community to get some ideas to assist you. Together, you should be able to come up with something that should be able to help you out!

Check these out:

[Next.js Discord](https://nextjs.org/discord)
[Next.js GitHub Discussions](https://github.com/vercel/next.js/discussions)
[Next.js Reddit](https://www.reddit.com/r/nextjs/)

Thank you so much. Please let me know if you have any more Vercel-related questions.

**Comment by Pieparker** [comment, marked as answer]
Hey Thamasky, check [this out](https://github.com/vercel/examples/tree/main/edge-functions/add-header)! Same principle should apply, using [this format](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) for strings.

**Reply by Thamasky** [reply]
Good to know, thanks!

**Reply by MerlinMason** [reply]
Hey there!

Just wondering if anyone has managed to get this working?
Looking at those links you provided I came up with the following:

```js
const response = NextResponse.next();

response.headers.set("Set-Cookie", [
    `a=1; Domain=${response.url}; SameSite=Strict`,
    `b=2; Domain=${response.url}; SameSite=Strict`,
]);
```

When running locally this creates two `set-cookie` headers, however when running on Vercel it creates none.

Is there something I'm missing?

**Reply by norio-club** [reply]
Same question! Please help

**Comment by volnei** [comment]
For me, still not working…

**Reply by amyegan** [reply]
Sorry you're still having trouble @volnei. Can you [start a new discussion](https://github.com/vercel/community/discussions/new) with details about what you have tried and how it is not working for you? 

If you can also share a deployment ID and [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) in your post, that will help us get a better view of what is going wrong.
